### PR TITLE
add first draft of my digest

### PR DESCRIPTION
**Pull request by donjar**

### CODE REVIEW

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
+---
+layout: post
+title: "Digest: Authentication with Ruby on Rails"
+date: 2017-02-16 12:00
+author: Herbert Ilhan Tanujaya
+categories: [Digest]
+---
+
+## Introduction
+Nowadays, authentication systems are present in nearly all websites (with at
```
Comment (jin) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: Nowadays, nearly all websites have an authentication system in place, using usernames and passwords.

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+least a username and password system in place). In this article, we will create
+a simple authentication system in the popular web application framework Ruby on
```
Comment (jin) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: in Ruby on Rails, a popular web application framework.

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+Rails. This article is mainly aimed at those who are just getting started with
+your first web application (but with some knowledge of Rails).
+
+When I first learned Rails (and web development in general), I wanted to create
```
Comment (jin) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: When I first learned Rails and web development in general,

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+a website for my organization. Since the website needs an authentication system,
+I searched for gems that help me. One of the most popular gems is
```
Comment (jin) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: One requirement for the website was to have an authentication system, so I searched for Ruby gems that could help with this.

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+[devise](https://github.com/plataformatec/devise); however, if you check the
```
Comment (jin) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: devise. However,

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+author: Herbert Ilhan Tanujaya
+categories: [Digest]
+---
+
+## Introduction
+Nowadays, authentication systems are present in nearly all websites (with at
+least a username and password system in place). In this article, we will create
+a simple authentication system in the popular web application framework Ruby on
+Rails. This article is mainly aimed at those who are just getting started with
+your first web application (but with some knowledge of Rails).
```
Comment (advaypal) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: their first web application(but have some knowledge of Rails).

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+When I first learned Rails (and web development in general), I wanted to create
+a website for my organization. Since the website needs an authentication system,
+I searched for gems that help me. One of the most popular gems is
+[devise](https://github.com/plataformatec/devise); however, if you check the
+README, you'll find out that if you are just getting started with Rails, it is
+recommended [not to use Devise](https://github.com/plataformatec/devise#starting-with-rails)!
+
+Based on my experiences, this article will walk through steps on creating a
```
Comment (advaypal) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: through the steps needed to create a

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+sufficiently secure authentication system in Ruby on Rails 5.0.1 on Ruby 2.4.0.
+
+## What to store in database?
+We will only set up a simple system with only usernames and passwords. Does that
+mean we store usernames and passwords directly in the database?
+
+It is totally not advisable to store plain passwords in the database. For one,
```
Comment (advaypal) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: remove totally

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+- computing the hash is quick and deterministic (meaning that hashing the same
+input always produces the same output)
+- given a hash, it is nearly impossible to find an input that hashes into it
+- a small change to the input makes a big change in the output
+
+For our purposes, we will be using [BCrypt](https://en.wikipedia.org/wiki/Bcrypt)
+that is conviniently bundled already in Ruby on Rails.
+
+There is another problem though - if we only store the password hash, an
+attacker can compute as many hashes as the attacker can, and store it in some
```
Comment (advaypal) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: as they can

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+Here's some explanation:
+- `attr_accessor :password` adds a "password" field in your User model that is
+not saved into the database.
+- The next two lines validate both username and password, by making sure they
+exist. Username is also ensured to be unique. Now you might realize that the
+username is already guaranteed in the database to be not null and unique; why
+bother doing it again in the model? Actually, it is enough to have it in either
+the model or the database (assuming the database only interacts with this app).
+I'm just used to including it in both.
```
Comment (advaypal) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: I'm just going to

File: _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+> u.authenticate('nush4ck3r5')
+=> false
+```
+So this means our user system is working. Yay!
+
+## What's next?
+For a simple authentication system that is only working in the model, this is
+sufficient. However, if you are going to create a full application with view
+and controller, you also need to add the login and logout functions, which
+needs more than this. There are many tutorials online - we might even add one
```
Comment (advaypal) [inline _posts/2017-02-18-digest-authentication-with-ruby-on-rails.md]: need

**Pull request merged**